### PR TITLE
Attempt to make DeferredK rerunnable

### PR DESCRIPTION
**Pull request by 1Jajen1**
Solves #1156 and maybe #1131 
Re-enable no-memoization law and fix deferredK to follow it (if used correctly)
Changes practically nothing from the public api.

I'll add some info to the docs later because if used wrong it might produce unexpected results.
Should the methods in DeferredK (well the public ones) have api docs (runnable ones) as well?

@pakoito I have skipped the invoke constructor for Wrapped because its syntax conflicts with the other one too much (the compiler was only able to infer which one to choose after passing more and more arguments which defeats the point of default argumens there), one should just use the other one or `.k()` instead. (And going from `() -> Deferred<A>` to `suspend () -> A` is trivial)

### CODE REVIEW

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt [hunk 1]
```diff
@@ -146,25 +143,25 @@ object MonadDeferLaws {
     df.flatMap { df }.flatMap { df }.equalUnderTheLaw(just(3), EQ) shouldBe true
   }
 
-  fun <F> MonadDefer<F>.stackSafetyOverRepeatedLeftBinds(iterations: Int = 5000, EQ: Eq<Kind<F, Int>>): Unit {
+  fun <F> MonadDefer<F>.stackSafetyOverRepeatedLeftBinds(iterations: Int = 50000, EQ: Eq<Kind<F, Int>>): Unit {
```
Comment (pakoito) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt]: 5k is enough to trigger a SO, 50k makes them too slow maybe.

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt [hunk 2]
```diff
@@ -127,13 +165,13 @@ fun <A> DeferredKOf<A>.unsafeRunSync(): A =
   runBlocking { await() }
 
 fun <A> DeferredKOf<A>.runAsync(cb: (Either<Throwable, A>) -> DeferredKOf<Unit>): DeferredK<Unit> =
-  DeferredK.invoke(scope(), Dispatchers.Unconfined, CoroutineStart.DEFAULT) {
+  DeferredK.invoke(fix().scope(), Dispatchers.Unconfined, CoroutineStart.DEFAULT) {
```
Comment (pakoito) [inline modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: Fix shouldn't be necessary here, `scope()` is defined as an alias to `fix().scope` atop of the file.

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt [hunk 3]
```diff
@@ -43,42 +82,41 @@ data class DeferredK<out A>(private val deferred: Deferred<A>, val scope: Corout
...
-
-    fun <A> failed(t: Throwable): DeferredK<A> =
-      CompletableDeferred<A>().apply { completeExceptionally(t) }.k()
+    operator fun <A> invoke(scope: CoroutineScope = GlobalScope, ctx: CoroutineContext = Dispatchers.Default, start: CoroutineStart = CoroutineStart.LAZY, f: suspend () -> A): DeferredK<A> =
+      Generated(ctx, start, scope, f)
 
     fun <A> raiseError(t: Throwable): DeferredK<A> =
       failed(t)
 
+    fun <A> failed(t: Throwable): DeferredK<A> =
```
Comment (pakoito) [inline modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: *squint* why is this here? why do we need an alias?

Comment (pakoito) [reply modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: I know I wrote it, I cannot remember why.

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt (hunk 1 above)
Comment (1Jajen1) [reply modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt]: Done

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt (hunk 2 above)
Comment (1Jajen1) [reply modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: :+1:

File: modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt (hunk 3 above)
Comment (1Jajen1) [reply modules/effects/arrow-effects-kotlinx-coroutines/src/main/kotlin/arrow/effects/DeferredK.kt]: Other than providing a more intuitive constructor for those that have never used raiseError I don't see the point either :) I'll remove it for now

**Comment by 1Jajen1** [comment]
It's not really required to distinguish between instantiation strategies (only for .k(), all others will work fine if a user does not use a deferred from outside that deferredK). So I think a section in the docs for DeferredK would be enough there. I can (and probably will in the next hour or so) add some kdoc to the api methods of DeferredK with examples. But except for .k() I don't think its worth mentioning retry-ability every time.

**Comment by pakoito** [comment]
Before merging I would hide the internal nodes behind a constructor and mark them as `internal data class`, else it's an API leak the users don't care about.

**Comment by 1Jajen1** [comment]
You mean marking Wrapped and Generated as Internal? They are already afaik. Changed them to data classes as well now. Do data classes override equals and stuff from their parent? o.O If so I may need to add equals/hashCode to those.

**Comment by pakoito** [comment]
Don't make them into `data`, you were right all along :D

**Comment by 1Jajen1** [comment]
Good thing git makes it nice and easy to revert stuff :)

**Comment by pakoito** [comment]
>Could you please add one more override of `invoke` that works as `.k()`, taking the scope too?

That and I'm good to merge :D

**Comment by raulraja** [comment]
@1Jajen1 yes ideally all methods are documented with at lest one example and a brief description of what they do. We consider this extremely important and are trying to enforce it lately in all PRs to ensure new users have a good experience when visiting the Arrow documentation. Each one of your methods then will have a dedicated page like this one http://arrow-docs.s3-website-us-east-1.amazonaws.com/docs/apidocs/arrow-effects/arrow.effects.typeclasses/-bracket/bracket-case.html

**Comment by 1Jajen1** [comment]
@raulraja Yeah I saw those, built it locally. The jekyll command took a while but worked out fine o.O
Will add it then. :+1: 

@pakoito So like an invoke that takes a Deferred? As in a convenience method over `defer { deferred.k() }` I'll do that, but I don't think the deferred can be supplied by a function, as the type signature of that will be very similar to the other one, thus when invoking it'll be annoying the user that it does not know if it has a `suspend () -> Deferred<A>` or a `() -> Deferred<A>`.

**Comment by pakoito** [comment]
The function signature is `operator fun <A> invoke(scope: CoroutineScope = GlobalScope, f: Deferred<A>): DeferredK<A>`, with a value instead of a function.

**Comment by 1Jajen1** [comment]
Added a lot of kdoc to almost all methods of DeferredK and the invoke constructor.

Btw. can you give generics a different color, like orange or something, in the api docs? Would make it so much more readable :)

**Comment by raulraja** [comment]
@1Jajen1 yes, feel free to create an issue and assign it to Israel Perez

**Comment by raulraja** [comment]
It's the number of libraries. As we move past 1.0 most of them will be in their own repository. Effects, streams, recursion schemes etc... we'll all be able to evolve differently and then the build times will be much smaller. In the mean-time, if you find areas where the build times can improve feel free to raise issues.

**Comment by 1Jajen1** [comment]
I'll look into speeding up gradle somehow and what exactly takes how much time... Don't know if I find stuff to improve, but worth trying :)

**Comment by 1Jajen1** [comment]
So the build itself consistently takes about 20 mins (15-18 on my laptop) with almost all time spent on tests (kapt is also pretty slow, but its nothing compared to the tests). 
Running `dokka :arrow-docs:runAnk` takes me about 3 mins, travis does not even get it done in 30 and my guess is that it's memory consumption. I frequently go up beyond 8-9 GB for that java task running it. (Which in turn slows down travis that does not have nearly as much memory available... Add another 10k files and let it run 2-3 hours and it'll probably throw a gc exception again)
So thats what needs to be dealt with. At that stage it's either dokka, ank or gradle. And considering memory consumption jumps up only when ank is processing it might still be it. I know it reports mem usage as ~1-2Gb all the time but something is obviously wrong :/

**Comment by raulraja** [comment]
I fixed ank memory problems in master a while back, did you merge since then?

**Comment by 1Jajen1** [comment]
Yes. It reports memory usage to be arount 1.5GB.
The process that runs ank (not confirmed, just the only process actually doing something when ank prints output) however consumes a ton of memory, way more than printed out (usually going to 6-8GB).

Some ideas I had, but could not test yet:
- The 1.5GB is inaccurate (which I don't think because it's read out correctly)
- The script engine runs in a seperate process and consumes a ton of memory which the parent just does not include (just an idea, don't know how it works)
- The ank plugin consumes a ton of memory. (Since this delegates to ank this seems unlikely)

**Comment by 1Jajen1** [comment]
I can confirm the reported memory usage to be the actual heap memory usage, but also that the process running ank is actually consuming up to 4-5 times that, which means stuff not on the heap is leaking. Although I have no idea what causes that.

File: modules/ank/arrow-ank/build.gradle [hunk 4]
```diff
@@ -1,9 +1,10 @@
 apply plugin: 'kotlin'
```
Comment (pakoito) [inline modules/ank/arrow-ank/build.gradle]: Is this a refactor or a merge? O.O

**Comment by 1Jajen1** [comment]
So I ended up not finding the reason why so much off-heap memory was used and resorted to rewriting ank to use `java.nio` instead of `java.io` (for the most part). I also changed and optimised a few things here and there. Most logic for processing input is still the same, just the way its invoked changed a bit.

It now consumes up to ~3GB memory at most. (Heap ~2GB, rest is probably memory mapped io)
It is also considerably faster (at least it feels faster, maybe that has to do with my laptop not freezing and putting stuff on swap to keep running o.O)

I ended up putting the changes on this branch to test it on travis as well, if this should be in a seperate pr please let me know and I'll revert that :) Either way please check if this code is correct or produces good ouput, I couldn't find differences from running ank on master but you never know :see_no_evil:

File: modules/ank/arrow-ank/build.gradle (hunk 4 above)
Comment (1Jajen1) [reply modules/ank/arrow-ank/build.gradle]: Too fast again :)
Explained this below.

File: modules/ank/arrow-ank/src/main/kotlin/arrow/ank/algebra.kt [hunk 5]
```diff
@@ -1,33 +1,34 @@
...
+import java.nio.file.Path
 
-fun createTarget(source: File, target: File): File =
-  createTargetImpl(source, target)
+interface AnkOps<F> {
+  fun MF(): MonadThrow<F>
 
-fun getFileCandidates(target: File): List<File> =
-  getFileCandidatesImpl(target)
+  fun createTargetDirectory(source: Path, target: Path): Kind<F, Try<Path>>
```
Comment (raulraja) [inline modules/ank/arrow-ank/src/main/kotlin/arrow/ank/algebra.kt]: There is no point on having concrete `Kind<F, Try<Path>` return type since they can just be `Kind<F, Path>` given you already have in scope a constrain on `MonadThrow<F>`. Why is the internal `Try` needed? You may just call `MF().raiseError` or use `MF().catch { ... }` if you just need to control synchronous error control. We can make the entire set of ops return `IO<A>` if you'd rather use an explicit concrete data type.

Comment (1Jajen1) [reply modules/ank/arrow-ank/src/main/kotlin/arrow/ank/algebra.kt]: Yeah, I kinda forgot when defining the algebra and then did not change it when implementing ...
Will do in a sec. Just removing a bunch of them should be fine since `MonadDefer.delay` also catches errors.
I had the type as `Kind<F, A>` because I was starting with only Monad as the constraint on F. That would make testing with something like an id interpreter nice. Now with MonadThrow this may not be valid anymore. (Although Try might be a good fit there)

**Comment by 1Jajen1** [comment]
Changed the signatures. (They really were kinda dumb...)

At least the whole build gets done in like 35 mins now. That leaves about 15 mins leeway for new tests and more kdoc to process before something has to be done about builds again, and if the repo gets split before that it likely won't be a problem ever :)

> The ank refactor is fine if it solves the memory issue for you but in whatever case we need to fix the nested Try signatures. You don't need those if you already have a constrain in MonadThrow<F>

Did you never get memory issues running it? o.O Maybe its a os specific thing then.

**Comment by raulraja** [comment]
No issues here and the latest in master is gonna push those to s3 instead of github pages because jekyll times out in github

**Comment by pakoito** [comment]
Soooooo...is anyone pushing this while the builds aren't broken and master is stable? :D

**Comment by 1Jajen1** [comment]
This should be good to go.
There is a bug in generating diagrams right now. Typeclasses which don't inherit from any other typeclass don't have a diagram atm. This is happening on master as well and I don't really know how and where they are generated or if this even may be intentional. For now I think merging this and creating a ticket should be fine.
